pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionCoordinatorClientImpl.java

```diff
@@ -18,42 +18,209 @@
+
+    private static final AtomicReferenceFieldUpdater<TransactionCoordinatorClientImpl, State> STATE_UPDATER =
+            AtomicReferenceFieldUpdater.newUpdater(TransactionCoordinatorClientImpl.class, State.class, "state");
+    private volatile State state = State.NONE;
+
+    public TransactionCoordinatorClientImpl(PulsarClient pulsarClient) {
+        this.pulsarClient = (PulsarClientImpl) pulsarClient;
+    }
+
+    @Override
+    public void start() throws TransactionCoordinatorClientException {
+        try {
+            startAsync().get();
+        } catch (Exception e) {
+            throw TransactionCoordinatorClientException.unwrap(e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> startAsync() {
+        if (STATE_UPDATER.compareAndSet(this, State.NONE, State.STARTING)) {
+            return pulsarClient.getLookup().getPartitionedTopicMetadata(TopicName.TRANSACTION_COORDINATOR_ASSIGN)
+                .thenAccept(partitionMeta -> {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Transaction meta store assign partition is {}.", partitionMeta.partitions);
+                    }
+                    if (partitionMeta.partitions > 0) {
+                        handlers = new TransactionMetaStoreHandler[partitionMeta.partitions];
+                        for (int i = 0; i < partitionMeta.partitions; i++) {
+                            TransactionMetaStoreHandler handler = new TransactionMetaStoreHandler(i, pulsarClient,
+                                    TopicName.TRANSACTION_COORDINATOR_ASSIGN.toString() + TopicName.PARTITIONED_TOPIC_SUFFIX + i);
+                            handlers[i] = handler;
+                            handlerMap.put(i, handler);
+                        }
+                    } else {
+                        handlers = new TransactionMetaStoreHandler[1];
+                        TransactionMetaStoreHandler handler = new TransactionMetaStoreHandler(0, pulsarClient,
+                                TopicName.TRANSACTION_COORDINATOR_ASSIGN.toString());
+                        handlers[0] = handler;
+                        handlerMap.put(0, handler);
+                    }
+
+                    STATE_UPDATER.set(TransactionCoordinatorClientImpl.this, State.READY);
+
+                });
+        } else {
+            return FutureUtil.failedFuture(new CoordinatorClientStateException("Can not start while current state is " + state));
+        }
+    }
+
+    @Override
+    public void close() throws TransactionCoordinatorClientException {
+        try {
+            closeAsync().get();
+        } catch (Exception e) {
+            throw TransactionCoordinatorClientException.unwrap(e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> closeAsync() {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        if (getState() == State.CLOSING || getState() == State.CLOSED) {
```

Comment:
@codelipenghui It seems the state can be transformed only:

* FROM `NONE` to `STARTING` in L73.
* FROM `STARTING` to `READY` in L95.

... never using `CLOSING` or `CLOSED`. It keeps even in the latest master. Any thoughts here?